### ISSUE TITLE
0.4:  Pkg.installed() does not work behind a firewall

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaLang/julia#15382]
I am setting up a lab right now which does not have outside access, so METADATA cannot be created locally.  I have all the packages installed locally in one of the LOAD_PATH locations, and almost all the packages work fine.  The exception is packages that call `Pkg.installed`, which tries to clone METADATA locally.  I think Compose.jl and Plots.jl are two culprits.

This seems like a bug: it shouldn't require remote access only to test if a package is installed.

**Comment by yuyichao** [comment]
Dup of https://github.com/JuliaLang/julia/issues/8679

**Comment by dlfivefifty** [comment]
Sorry about that, I searched for Pkg.installed and didn't see anything.

**Comment by wildart** [comment]
@dlfivefifty You need to copy METADATA on each machine, because it is the only source of information for a package versioning and dependency resolution.

**Comment by dlfivefifty** [comment]
But `using PKG_NAME` works successfully with packages shared in LOAD_PATH.  If its possible to load a package, it should be possible to see if its installed.

**Comment by wildart** [comment]
There are many things in `Pkg` module which depend on presence of METADATA, and `Pkg` module has nothing to do with `using` command that loads particular package sources.

**Comment by dlfivefifty** [comment]
So what’s the correct way to tell if using is going to be successful, then?

**Comment by wildart** [comment]
Any package in `Pkg.dir()` or `LOAD_PATH` is available for `using`. As for installed packages, situation a bit different. Package is qualified as "installed" if it was explicitly installed with a package manager `Pkg.add` command. Installed packages are shown in "required packages" section of `Pkg.status`, dependent packages and manually installed will be listed in "additional packages" section.

Bottom line: if you do not use package manager then there is no way to tell which packages are installed in a system.

**Comment by dlfivefifty** [comment]
Is there a reason that it can't use the METADATA in LOAD_PATH?

**Comment by yuyichao** [comment]
> Is there a reason that it can't use the `METADATA` in `LOAD_PATH`?

Well `METADATA` won't be particularly useful but I don't see any reason we can't use the metadata that come with a package for read only package manager operations. Especially since what another package usually want to check is "whether package Foo is available" and not "whether package Foo is installed by the package manager".

This is essentially what #8679 is about. For this particular case, since `Pkg.installed()` returns a version that may not be easy to figure out for system wise installed package without a `.git` directory, it might be necessary to let the package manager read the version from an optional file in the package root that can be generated by the system wise package manager.

**Comment by tkelman** [comment]
> might be necessary to let the package manager read the version from an optional file in the package root that can be generated by the system wise package manager.

add to the list for https://github.com/JuliaLang/julia/issues/11955

**Comment by wildart** [comment]
> Well METADATA won't be particularly useful but I don't see any reason we can't use the metadata that come with a package for read only package manager operations. 

That is the point. Julia package system is based on centralized metadata storage and package location (with some exceptions). Even though individual packages carry versioning and dependency information, the package manager uses this info only when it is signed and incorporated in to METADATA database.
Yes, Julia allows to sideload packages from `LOAD_PATH`, but the package manager is build around  METADATA storage which resides in predetermined location.

BTW, how about conflict resolution, when the same package is in multiple locations? Without keeping centralized package storage or local copy of all dependencies, it will be very hard to find out which version to load.

**Comment by yuyichao** [comment]
> Julia package system is based on centralized metadata storage and package location

The package manager is, the package loader is not. The issue is basically about doing read only package operations on packages that can be found by the loader.

> Without keeping centralized package storage or local copy of all dependencies, it will be very hard to find out which version to load.

Same with the above, this is not a issue at all. The loading find the first one in the order specified in various variables and the package info query just need to follow the same order.

**Comment by yuyichao** [comment]
Another way to describe this is that there should be a function to check the availability of a package and where it is/would be loaded from. It does not necessarily has to be extending existing package manager functions but in cases when there's little unambiguity I think it is better to just extend the current functions (otherwise, packages using the package manager function would need to change to the new functions to support LOAD_PATH packages and this would effectively deprecate the package manager one).
